widget_driver/doc/widget_drivers.md

```diff
@@ -0,0 +1,142 @@
+
+```dart
+part 'counter_widget_driver.g.dart';
+
+@Driver()
+class CounterWidgetDriver extends WidgetDriver {
+  final CounterService _counterService;
+  final Localization _localization;
+  StreamSubscription? _subscription;
+
+  CounterWidgetDriver(
+    BuildContext context, {
+    CounterService? counterService,
+  })  : _counterService = counterService ?? GetIt.I.get<CounterService>(),
+        _localization = context.read<Localization>(),
+        super(context) {
+
+    _subscription = _counterService.valueStream.listen((_) {
+      notifyWidget();
+    });
+
+  }
+
+  @DriverProperty('The title of the counter')
+  String get counterTitle => _localization.counterTitle;
+
+  @DriverProperty(1)
+  int get value => _counterService.value;
+
+  @DriverAction()
+  void increment() {
+    _counterService.increment();
+  }
+
+  @DriverAction()
+  void decrement() {
+    _counterService.decrement();
+  }
+
+  @override
+  void dispose() {
+    _subscription?.cancel();
+    super.dispose();
+  }
+}
+```
+
+There are three important concepts here.
+
+1. The dependencies are loaded/crated/resolved in the constructor of the `driver`.
+1. All the properties/methods which the driver wants to expose to the widgets needs to be annotated and given default values.
+1. Whenever important data has changed in the `driver` you need to call `notifyWidget()`
+
+**Let's start talking about the dependencies.**  
+Any type of dependencies which your `driver` needs to be able to give your widget the correct data has to be loaded/crated/resolved in the constructor of the `driver`.
+
+You get two option here for how to resolve these dependencies. Either you can grab them out of the `BuildContext` (for example if you are using the `Provider` package). Or you can use some DI package like `get_it` and then just grab the dependencies from the DI container.
+
+**Now what about those annotations?**  
+They are needed for the `testDrivers` to be generated correctly. So for all properties and methods which you expose to the widget, you will need to add these annotations.
+
+For properties you add `@DriverProperty({default_value})`.  
+The `default_value` should be the default value which you want to use when this widget is being created by other widgets under test.
+
+For methods you add `@DriverAction()`.  
+This is needed so that any potential parent widget when tested can call these methods without crashing.
+
+After you are done creating your `driver` you need to run the generator so that the `testDrivers` gets created. Read more [here](code_generation.md) about running the generators.
+
+**Updating the widget.**  
+So now you have your dependencies and they have emitted some new state and you want to update your widget. But how?!
+
+Easy, you just call `notifyWidget()`. This will force your widget to call its `build` method again and update. Then it will just read the latest values from your `driver`.
```

Comment:
```suggestion
That's easy: Just call `notifyWidget()`. This will force your widget to call its `build` method again and redraw using the latest values from your `driver`.
```